pkg/chunk/cache/redis_cache.go

```diff
@@ -5,48 +5,91 @@ import (
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	instr "github.com/weaveworks/common/instrument"
 
 	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
 )
 
 // RedisCache type caches chunks in redis
 type RedisCache struct {
-	name   string
-	redis  *RedisClient
-	logger log.Logger
+	name            string
+	redis           *RedisClient
+	logger          log.Logger
+	requestDuration observableVecCollector
 }
 
 // NewRedisCache creates a new RedisCache
-func NewRedisCache(name string, redisClient *RedisClient, logger log.Logger) *RedisCache {
+func NewRedisCache(name string, redisClient *RedisClient, reg prometheus.Registerer, logger log.Logger) *RedisCache {
 	util_log.WarnExperimentalUse("Redis cache")
 	cache := &RedisCache{
 		name:   name,
 		redis:  redisClient,
 		logger: logger,
+		requestDuration: observableVecCollector{
+			v: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+				Namespace:   "cortex",
+				Name:        "rediscache_request_duration_seconds",
+				Help:        "Total time spent in seconds doing Redis requests.",
+				Buckets:     prometheus.ExponentialBuckets(0.000016, 4, 8),
+				ConstLabels: prometheus.Labels{"name": name},
+			}, []string{"method", "status_code"}),
+		},
 	}
 	if err := cache.redis.Ping(context.Background()); err != nil {
 		level.Error(logger).Log("msg", "error connecting to redis", "name", name, "err", err)
 	}
 	return cache
 }
 
+func redisStatusCode(err error) string {
+	// TODO: Figure out if there are more error types returned by Redis
+	switch err {
+	case nil:
+		return "200"
+	default:
+		return "500"
+	}
+}
+
 // Fetch gets keys from the cache. The keys that are found must be in the order of the keys requested.
 func (c *RedisCache) Fetch(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string) {
-	data, err := c.redis.MGet(ctx, keys)
+	const method = "RedisCache.MGet"
+	var items [][]byte
+	// Run a tracked request, using c.requestDuration to monitor requests.
+	err := instr.CollectedRequest(ctx, method, c.requestDuration, redisStatusCode, func(innerCtx context.Context) error {
+		log, _ := spanlogger.New(innerCtx, method)
+		defer log.Finish()
+		log.LogFields(otlog.Int("keys requested", len(keys)))
+
+		var err error
+		items, err = c.redis.MGet(ctx, keys)
```

Comment:
Should we use `innerCtx` here? Perhaps it would be even better to call parameter in this function just `ctx`, to avoid mixing two contexts.

Reply:
Agreed it should be, but we could merge this then fix it in post?
I looked at the Redis code; it doesn't seem to do anything with the context that would break with this code.

Reply:
> Agreed it should be, but we could merge this then fix it in post?

Fine with me.

Reply:
@pstibrany thanks, renamed `innerCtx` to `ctx`.